apps/dashboard/lib/auth/get-auth.ts

```diff
@@ -1,64 +1,133 @@
+import { cache } from "react";
+import { env } from "../env";
 import { getCookie } from "./cookies";
 import { auth } from "./server";
-import { UNKEY_SESSION_COOKIE } from "./types";
+import { UNKEY_REFRESH_TOKEN, UNKEY_SESSION_COOKIE } from "./types";
 
 type GetAuthResult = {
   userId: string | null;
   orgId: string | null;
+  role: string | null;
 };
 
-export async function getAuth(_req?: Request): Promise<GetAuthResult> {
+// Cache session validation based on session token
+export const validateSessionCached = cache(
+  async (
+    sessionToken: string | null,
+  ): Promise<{
+    isValid: boolean;
+    shouldRefresh: boolean;
+    userId?: string;
+    orgId?: string | null;
+    role?: string | null;
+  }> => {
+    if (!sessionToken) {
+      return { isValid: false, shouldRefresh: false };
+    }
+
+    try {
+      return await auth.validateSession(sessionToken);
+    } catch (error) {
+      console.error("Session validation error:", error);
+      return { isValid: false, shouldRefresh: false };
+    }
+  },
+);
+
+// Global mutex objects to prevent concurrent refresh operations
+let refreshInProgress = false;
+let refreshPromise: Promise<GetAuthResult> | null = null;
```

Comment:
This being on fluid compute... could be a problem. In a edge case scenario.

Scenario: This global context could run on the same VM for two users and then request a new access / refresh token you potentially could end up giving a new access token / refresh token to the wrong user. Which in turn could potentially expose us to wrong user data.

Reply:
Going to put my PR in draft to address this since I pushed the other fixes

Reply:
addressed in latest commits: updated mutex to be per-user, using a map and checking the session validation result with the refresh token in the map to ensure it goes back to the right user